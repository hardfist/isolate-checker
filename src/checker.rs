use miette::{Report, Result};
use std::{cell::RefCell, sync::Arc};

use crate::{
    ast::Ast,
    hir_ctx::HirCtx,
    infer::{InferContext, TypeInference},
};

pub struct ModuleChecker {
    errors: RefCell<Vec<Report>>,
    code: Arc<String>,
    ast: Ast,
}
impl ModuleChecker {
    pub fn new(code: Arc<String>) -> Result<ModuleChecker> {
        let ast: Ast = Ast::new_from(code.clone())?;
        Ok(Self {
            errors: RefCell::new(vec![]),
            code,
            ast,
        })
    }
}
impl ModuleChecker {
    pub fn extends_errors(&self, errors: Vec<Report>) {
        self.errors.borrow_mut().extend(errors);
    }
    pub fn check(&self) {
        let mut errors = Vec::new();
        let hir_ctx = HirCtx::new(&self.ast, &mut errors);
        let mut infer = TypeInference::default();
        let infer_ctx = InferContext::new(&hir_ctx);

        for item in hir_ctx.ast.items() {
            infer.infer_item(&infer_ctx, item);
        }

        errors.append(&mut infer.reports);
        self.extends_errors(errors);

        let error_msg = self.emit_error();
        self.emit_type(&infer);
        println!("{:?}", error_msg);
    }
    pub fn emit_type(&self, ctx: &TypeInference) {
        for (node_id, ty) in &ctx.typemap {
            dbg!(node_id, ty);
        }
    }
    pub fn emit_error(&self) -> String {
        let report_handler = miette::GraphicalReportHandler::new();
        let error_msg: Vec<String> = self
            .errors
            .take()
            .into_iter()
            .map(|err| {
                let err = err.with_source_code(self.code.clone());
                let mut output = String::new();
                report_handler
                    .render_report(&mut output, err.as_ref())
                    .unwrap();
                output
            })
            .collect();
        error_msg.join("\n")
    }
}
