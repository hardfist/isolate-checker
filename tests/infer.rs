use isolate_checker::checker::ModuleChecker;
use miette::IntoDiagnostic;
use std::fs;
use std::path::Path;
use std::sync::Arc;

#[test]
pub fn run_test() -> miette::Result<()> {
    let root = std::env!("CARGO_MANIFEST_DIR");
    let fixtures_dir = Path::new(root).join("./tests/fixtures");
    let cases = std::fs::read_dir(fixtures_dir).expect("get fixture failed");
    for case in cases.into_iter() {
        let case = case.into_diagnostic()?;
        assert!(case.path().extension().map_or(false, |x| dbg!(x).eq("ts")));
        let content = fs::read_to_string(case.path()).into_diagnostic()?;
        let mut checker = ModuleChecker::new(Arc::new(content))?;
        checker.check();
        let error_msg = checker.emit_error();
        insta::assert_snapshot!("basic.err", error_msg);
        
    }
    Ok(())
}
