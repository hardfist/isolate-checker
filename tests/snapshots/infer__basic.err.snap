---
source: tests/infer.rs
expression: error_msg
---

